### PR TITLE
plo: add uart-imx6ull

### PR DESCRIPTION
**Pull request by HBuczynski**
## Description

- add circular buffer
- add uart-imx6ull

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (imx6ull).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: devices/uart-imx6ull/uart.c [hunk 1]
```diff
@@ -0,0 +1,405 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * i.MX 6ULL Serial driver
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Hubert Buczynski
```
Comment (agkaminski) [inline devices/uart-imx6ull/uart.c:9]: It was in part copied from phoenix-rtos-devices, wasn't it? Keep the original authors and dates

Comment (HBuczynski) [reply devices/uart-imx6ull/uart.c]: Only some small parts are derived form phoenix-rtos-devices, the majority of the code differs so it is the reason why I skipped the other authors.
Nevertheless, I have updated the previous authors ;)

File: devices/uart-imx6ull/uart.c [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+
+	md = MODULE_CLK;
+	in = 16 * baud;
+
+	div = 0;
+	while (md % 2 == 0 && in % 2 == 0) {
+		md = md / 2;
+		md = md / 2;
+		div++;
+	}
```
Comment (kemonats) [inline devices/uart-imx6ull/uart.c:174]: Is this code correct?
`in % 2 == 0` is always true, earlier `in = 16 * baud`, in loop `in` is not changed
twice `md = md / 2` and not `md = md / 4`

Comment (HBuczynski) [reply devices/uart-imx6ull/uart.c:174]: I borrowed this function from phoenix-rtos-devices and I didn't deep into details of implementation, assuming that it works fine.
https://github.com/phoenix-rtos/phoenix-rtos-devices/blob/2b8ed8fc02895fd316cc8bab8d98438fa70a9970/tty/imx6ull-uart/imx6ull-uart.c#L225-L273

However, I agree with your suggestions that some of the operations are unnecessary.

File: devices/uart-imx6ull/uart.c [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+} uart_common;
+
+
+static inline void uart_rxData(uart_t *uart)
+{
+	char c;
+
+	/* Keep getting data until fifo is not empty */
+	while (*(uart->base + usr2) & 0x1) {
+		c = *(uart->base + urxd) & 0xff;
```
Comment (kemonats) [inline devices/uart-imx6ull/uart.c:100]:
```suggestion
		c = *(uart->base + urxd);
```
Mask is not needed, `c` is type `char`.

File: devices/uart-imx6ull/uart.c [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+	return 0;
+}
+
+
+static void uart_padsInit(unsigned int minor)
+{
+	int i;
+
+	for (i = 0; i < 4; i++)
+		imx6ull_setIOmux(uart_mux[(minor)][i].ioctl, 0, uart_mux[(minor)][i].val);
```
Comment (kemonats) [inline devices/uart-imx6ull/uart.c:152]:
```suggestion
		imx6ull_setIOmux(uart_mux[minor][i].ioctl, 0, uart_mux[minor][i].val);
```
Unneccesary brackets.

File: devices/uart-imx6ull/uart.c [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+
+	for (i = 0; i < 2; i++)
+		imx6ull_setIOisel(uart_isel[(minor)][i].ioctl, uart_isel[(minor)][i].val);
```
Comment (kemonats) [inline devices/uart-imx6ull/uart.c:155]:
```suggestion
		imx6ull_setIOisel(uart_isel[minor][i].ioctl, uart_isel[minor][i].val);
```
Unneccesary brackets.

File: devices/uart-imx6ull/uart.c (hunk 2 above)
Comment (kemonats) [reply devices/uart-imx6ull/uart.c:174]: Shouldn't there be something like this?
```
	while (md % 2 == 0 && in % 2 == 0) {
		md = md / 2;
		in = in / 2;
		div++;
	}
```

Comment (HBuczynski) [reply devices/uart-imx6ull/uart.c]: Yes, you have right. Nevertheless, I have changed implementation of baud rate calculation.

File: devices/uart-imx6ull/uart.c (hunk 4 above)
Comment (HBuczynski) [reply devices/uart-imx6ull/uart.c]: Thx. Fixed.

File: devices/uart-imx6ull/uart.c (hunk 5 above)
Comment (HBuczynski) [reply devices/uart-imx6ull/uart.c]: Thx. Fixed.

File: devices/uart-imx6ull/uart.c [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+		den /= max;
+
+		if (!num)
+			num = 1;
+
+		if (!den)
+			den = 1;
+	}
+
+	calcBaudrate = (UART_REF_CLK >> 4) * den / num;
```
Comment (kemonats) [inline devices/uart-imx6ull/uart.c:193]:
```suggestion
	calcBaudrate = ((UART_REF_CLK >> 4) * den) / num;
```
safer

File: devices/uart-imx6ull/uart.c [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+	diff = (calcBaudrate >= baudrate) ? (calcBaudrate - baudrate) : (baudrate - calcBaudrate);
+
+	if ((diff * 100 / baudrate) > 3)
```
Comment (kemonats) [inline devices/uart-imx6ull/uart.c:196]:
```suggestion
	if (((diff * 100) / baudrate) > 3)
```
safer

**Review by agkaminski — APPROVED** [review]

**Pull request merged**